### ISSUE TITLE
Jolt RID leak on shutdown?

### ISSUE THREAD

**Issue by Gamepro5** [godot-jolt/godot-jolt#520]
![image](https://github.com/godot-jolt/godot-jolt/assets/23440295/e70a8cec-c20b-4d99-b757-255ac9e3b77b)
 Here is the errors reported. I am not 100% sure what an RID is, but there is a possibility that some of them are because I add physics object references in scripts. The thing I don't understand is that this isn't constant, even without scripts. This error was from my own project, and I can't replicate it with a super basic project. I'm wondering what this error means so I can try to look for any causes in my own project, if I can't find any, I know there's a bigger issue here.

**Comment by mihe** [comment]
tl;dr: You likely have orphaned nodes that need to be freed. Searching for "godot orphaned nodes" should yield some useful tips for how to deal with this.

You can think of Godot as having a front-end and a back-end. The nodes that you interact with, like `RigidBody3D` and `CharacterBody3D`, are the front-end and typically deal with just setting up the properties, methods and maybe some high-level logic. The nodes then delegate a lot of the actual work to the back-end, in this case known as the "[physics server](https://docs.godotengine.org/en/stable/classes/class_physicsserver3d.html)", which is the interface that Godot Jolt implements.

The way these two communicate is by the nodes allocating some type of resource from the server, in this case something like a body or a shape, which is then referenced using a `RID`, or resource ID. Since the server has no idea about the actual `RigidBody3D` it also has no notion of the scene tree, nor any notion about when the node is created/destroyed. So it's the responsibility of the node to free/release/deallocate the resource that the `RID` points to when the node itself is freed.

In a normal game/application flow all of the nodes would belong to the scene tree, which would lead them all to be freed whenever the scene tree itself is freed. In your case, however, it's looking like you probably have some nodes that live outside the scene tree (also known as "orphaned nodes") and thus aren't freed when the game/application is shutting down, leading to all of the `RID`s that they're associated with also leaking.

I haven't had to deal with this myself, but from looking around it seems like there are some tools available in Godot to help figure this out, like the `Node.print_orphan_nodes()` static method, or running the game/application with `--verbose`, which will print any leaked/orphaned node instances on shutdown.

**Comment by Gamepro5** [comment]
That was indeed the problem. The --verbose isn't verbose enough. Print orphan nodes helped me quite a few times. The error was that the entire map (with several collision shapes) simply were removed from the tree and not properly deleted. Sorry for wasting your time, I just saw "report bug" and my monkey brain went off.
